app/features-json/user_json_controller.rb

```diff
@@ -0,0 +1,58 @@
+require_relative "../services/user_service"
+require_relative "api_controller"
+require_relative "json_params"
+
+module FastlaneCI
+  # Controller responsible for handling users
+  class UserJSONController < APIController
+    disable(:authentication)
+
+    HOME = "/api/user"
+
+    post HOME.to_s do
+      email = params[:email]
+
+      if email.nil?
+        # User didn't provide a custom email address
+        # fetch it based on the API token instead
+        github_client = Octokit::Client.new(access_token: params[:github_token])
+
+        begin
+          # Note: This fails if the user.email scope is missing from token
+          email = github_client.emails.find(&:primary).email
+        rescue Octokit::NotFound
+          json_error!(
+            error_message: "Provided API token needs user email scope",
+            error_key: "User.Token.MissingEmailScope",
+            error_code: 400
+          )
+        rescue Octokit::Unauthorized
+          json_error!(
+            error_message: "Provided API token is invalid",
+            error_key: "User.Token.Invalid",
+            error_code: 403
+          )
+        end
+      end
+
+      user = Services.user_service.create_user!(
+        email: email,
+        password: params[:password]
+      )
+
+      if user
+        Services.user_service.create_provider_credential!(
+          user_id: user.id,
+          email: user.email,
+          api_token: params[:github_token]
+        )
+        return json({ status: :success })
+      else
+        json_error!(
```

Comment:
could we single out the error where the user already exists? It would provide better messaging in the client.

Reply:
Yeah, in this case we don't get any error information - as the `create_user!` method doesn't expose the error type as of now. There are probably gonna be other error types in the future. I'd be in favor of keeping the generic message for now, until we spend more time refining what kind of error messages we want, and how we want to handle them (e.g. missing config file, no internet connection, etc.) Right now we catch some of them, but most exceptions are just surfaced all the way up, and therefore not returning an actual JSON object. 

We can't just expose the stack trace and exception AFAIK as it potentially contains sensitive information.